File: examples/java/matrices/src/Matrix.java
import java.util.ArrayList;
import java.util.List;

public class Matrix {
    private final double[][] values;

    public Matrix(double[][] values) {
        this.values = values;
    }

    public double sum() {
        var result = 0.0;
        for (double[] value : values) {
            result += addRow(value);
        }
        return result;
    }

    public double sumParallel() {
        List<MathThread> mathThread = new ArrayList<>();

        // se crean los threads, acá se podrían startear
        for (double[] value : values) {
            mathThread.add(new MathThread(() -> addRow(value)));
        }

        // se startean todos
        mathThread.forEach(Thread::start);

        // acá se calcula el valor
        mathThread.forEach(m -> {
            try {
                m.join();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        });

        // acá se pasan todos los getValue a una colección y luego se suma
        return mathThread.stream().mapToDouble(MathThread::getValue).sum();
    }

    public Matrix addSerial(Matrix other) {
        double[][] result = new double[values.length][];
        for (int i = 0; i < values.length; i++) {
            int cols = values[i].length;
            var row = new double[cols];
            for (int j = 0; j < cols; j++) {
                row[j] = values[i][j] + other.values[i][j];
            }
            result[i] = row;
        }
        return new Matrix(result);
    }

    public Matrix addParallel(Matrix other) {
        throw new RuntimeException("Implement me!");
    }

    private double addRow(double[] value) {
        var result = 0.0;
        for (double v : value) {
            result += v;
        }
        return result;
    }


}

File: examples/rust/matrices/src/matrix.rs
use std::thread;

#[derive(Debug, Clone)]
pub struct Matrix(pub Vec<Vec<f64>>);

impl Matrix {
    pub fn rows(&self) -> usize {
        self.0.len()
    }
    pub fn columns(&self) -> usize {
        self.0[0].len()
    }

    pub fn sum_serial(&self) -> f64 {
        let mut result = 0.0;
        for row in &self.0 {
            result += self.add_row(row);
        }
        return result;
    }

    pub fn sum_parallel(&self) -> f64 {
        // el scope lo que hace es crear todos los threads
        // dentro del scope, entonces cuando termina el scope
        // está garantizado que ya terminaron
        thread::scope(|s| {
            let mut threads = Vec::new();
            for row in &self.0 {
                // con el move, le pasas el ownership del row a la función de abajo
                threads.push(s.spawn(move || self.add_row(&row)));
            }
            let mut result = 0.0;
            for thread in threads {
                result += thread.join().unwrap();
            }
            return result;
        })
    }

    fn add_row(&self, row: &Vec<f64>) -> f64 {
        let mut result = 0.0;
        for v in row {
            result += v;
        }
        return result;
    }

    pub fn add_serial(&self, other: &Matrix) -> Matrix {
        let rows = self.rows();
        let cols = self.columns();
        let result = (0..rows)
            .map(|i| (0..cols).map(|j| self.0[i][j] + other.0[i][j]).collect())
            .collect();
        Matrix(result)
    }

    pub fn add_parallel(&self, _: &Matrix) -> Matrix {
        todo!("Implement me!")
    }
}
